### PR TITLE
fix(server): prevent open redirect via forged Origin header on login

### PR DESCRIPTION
**Pull request by subrata71**
## Summary

Fixes an open redirect vulnerability (APP-15347) where an attacker could forge the `Origin` HTTP header on `POST /api/v1/login` to redirect authenticated users to an external domain.

**Root cause:** `isSafeRedirectUrl()` used the client-supplied `Origin` header as both the redirect URL builder and the trust anchor. A forged `Origin: https://evil.com` passed validation because the redirect host trivially matched the Origin host (`evil.com == evil.com`).

**Fix:** Introduces `getTrustedOrigin()` which cross-checks the `Origin` header against the request's `Host` / `X-Forwarded-Host` (set by the reverse proxy, not the client). All four call sites that previously read Origin directly now use this helper:

- `isSafeRedirectUrl()` — no longer uses forged Origin as the validation baseline
- `fulfillRedirectUrl()` — no longer prepends forged Origin to relative redirect paths
- `sanitizeRedirectUrl()` — no longer uses forged Origin in the fallback URL
- `getRedirectUrl()` (fork-app path) — no longer uses forged Origin for URL construction

When no `Host` / `X-Forwarded-Host` is available (unusual environments without proxy headers), Origin is still trusted to preserve backward compatibility.

## Test plan

- [x] Added unit test `testForgedOriginHeaderIsBlockedWhenHostDiffers` — confirms forged Origin is rejected and sanitized URL does not contain the attacker domain
- [x] All 69 existing `RedirectHelperOpenRedirectTest` tests pass (0 failures, 0 errors)

## Linear ticket

https://linear.app/appsmith/issue/APP-15347/security-low-open-redirect-via-improper-validation-of-origin-header-on

## Summary by CodeRabbit

* **Bug Fixes**
  * Strengthened open-redirect protection by only trusting the `Origin` header when it matches the inbound request host (including effective port), derived from `X-Forwarded-Host`/`Host` (and forwarded port when present).
  * Updated redirect validation and fallback URL generation to prevent forged origin values from being used or leaked.

* **Tests**
  * Added open-redirect regression coverage for forged `Origin` handling, proxy-style `X-Forwarded-Host` behavior, and IPv4/IPv6 same-host different-port edge cases.

## Automation
/ok-to-test tags="@tag.All"

> [!TIP]
> 🟢 🟢 🟢 All cypress tests have passed! 🎉 🎉 🎉
> Workflow run: <https://github.com/appsmithorg/appsmith/actions/runs/29393469419>
> Commit: fea4a4aa6f6eba5d1f3a5d88173b5db49f8b3bbb
> <a href="https://internal.appsmith.com/app/cypress-dashboard/rundetails-65890b3c81d7400d08fa9ee5?branch=master&workflowId=29393469419&attempt=1" target="_blank">Cypress dashboard</a>.
> Tags: `@tag.All`
> Spec:
> <hr>Wed, 15 Jul 2026 07:24:36 UTC

### CODE REVIEW

File: app/server/appsmith-server/src/test/java/com/appsmith/server/helpers/RedirectHelperOpenRedirectTest.java [hunk 1]
```diff
@@ -500,4 +500,33 @@ void testSanitizeBlocksWhitespacePrefixedUrl() {
...
+
+    @Test
+    void testIPv6OriginMatchesRequestHost() {
+        HttpHeaders headers = new HttpHeaders();
+        headers.setOrigin("http://[::1]:8080");
+        headers.set("X-Forwarded-Host", "[::1]:8080");
+        assertTrue(
+                RedirectHelper.isSafeRedirectUrl("http://[::1]:8080/applications", headers),
+                "IPv6 Origin matching X-Forwarded-Host must be accepted");
+    }
```
[the remark this answers is not in the document]
Comment (subrata71) [reply app/server/appsmith-server/src/test/java/com/appsmith/server/helpers/RedirectHelperOpenRedirectTest.java:629]: Let's live without it for now.

**Review by wyattwalter — APPROVED** [review]
